compiler/rustc_mir/src/dataflow/impls/available_locals.rs

```diff
@@ -0,0 +1,338 @@
+use std::{fmt, hash::BuildHasherDefault};
+
+use rustc_data_structures::fx::FxIndexMap;
+use rustc_index::{bit_set::BitSet, vec::Idx};
+use rustc_middle::{
+    mir::{self, Body, Local, Location},
+    mir::{visit::Visitor, BorrowKind, Operand, Place, Rvalue, Statement, StatementKind},
+};
+use smallvec::SmallVec;
+
+use crate::dataflow::{
+    fmt::DebugWithAdapter, fmt::DebugWithContext, lattice, lattice::Dual, AnalysisDomain, Forward,
+    GenKill, GenKillAnalysis,
+};
+
+/// Dataflow analysis for availability of locals. A local is available at a given program point,
+/// if the value of the local can freely be used at the given program point.
+/// Consider the following example:
+/// ```
+/// _1 = 4;
+/// _2 = &_1;
+/// _3 = *_2
+/// ```
+/// In the above example, `_2` is available at the third statement, so the statement can be
+/// simplified to `_3 = _1`.
+/// In general, an available local can be used freely on any path from the definition of `_2` to
+/// statement `s`, if `_2` and its value is guaranteed to not be changed on all paths.
```

Comment:
What's the difference between this and reaching definitions? It doesn't really seem like the best idea to introduce an ad-hoc analysis for a single minor optimization instead of using a well-known analysis that may have other uses in the future.

Reply:
I may be a bit off on the terminology here. What I implemented sounds like reaching definitions, yeah. Did I miss an implementation already in rustc?

"available expression analysis" is often the term I see used in compiler theory, which refers to an analysis that can answer if an expression can be reused at a point since it is not modified along the way from its definition.

I named the pass I implemented "available locals" since it tracks locals, not expressions. But I could probably more precisely name it reaching definitions.

I agree that a new dataflow analysis solely for this one optimization is a bit overkill, but I think the analysis in general can be applied in a lot of coming or existing passes.